src/MusicPlaylistSecondary.java

```diff
@@ -1,4 +1,75 @@
+import components.map.Map;
 
-public class MusicPlaylistSecondary {
+/**
+ * Layered implementations of secondary methods for {@code MusicPlaylist}.
+ */
+public abstract class MusicPlaylistSecondary implements MusicPlaylist {
+
+    @Override
+    public final void shuffle() {
+        Map.Pair<String, String> removed = this.removeAny();
+        this.add(removed.value(), removed.key());
+    };
+
+    @Override
+    public final String artist(String song) {
+        Map.Pair<String, String> removed = this.remove(song);
+        String artist = removed.key();
+        this.add(removed.value(), removed.key());
+        return artist;
+    };
+
+    @Override
+    public final String upNext() {
+        Map.Pair<String, String> removed = this.removeAny();
+        String song = removed.value();
+        this.add(removed.value(), removed.key());
+        return song;
+    };
+
+    @Override
+    public final String toString() {
```

Comment:
I changed my mind about this. `toString()` is used by print, so any time someone tries to print your object it will cause a crash. You can either leave this unimplemented or you can print out an array of songs.